src/zcash/Note.hpp

```diff
@@ -40,24 +42,51 @@ class SproutNote : public BaseNote {
     uint256 nullifier(const SproutSpendingKey& a_sk) const;
 };
 
+inline bool plaintext_version_is_valid(const Consensus::Params& params, int height, unsigned char leadByte) {
+    int canopyActivationHeight = params.vUpgrades[Consensus::UPGRADE_CANOPY].nActivationHeight;
+
+    if (height < canopyActivationHeight && leadByte != 0x01) {
+        // non-0x01 received before Canopy activation height
+        return false;
+    }
+    if (height >= canopyActivationHeight
+        && height < canopyActivationHeight + ZIP212_GRACE_PERIOD
+        && leadByte != 0x01
+        && leadByte != 0x02)
+    {
+        // non-{0x01,0x02} received after Canopy activation and before grace period has elapsed
+        return false;
+    }
+    if (height >= canopyActivationHeight + ZIP212_GRACE_PERIOD && leadByte != 0x02) {
+        // non-0x02 received past (Canopy activation height + grace period)
+        return false;
+    }
+    return true;
+};
 
 class SaplingNote : public BaseNote {
+private:
+    uint256 rseed;
+    friend class SaplingNotePlaintext;
+    unsigned char leadByte;
```

Comment:
Storing a `leadByte` in the note itself is semantically invalid, and prone to error. The `leadByte` field of the note plaintext indicates semantic information about the note plaintext, not the note (i.e. it could indicate something about the memo independently from the note). As currently implemented (e.g. in `SaplingNote::rcm`), if `SaplingNote::leadByte != 0x02` then the note is treated as a "version 1" note, which immediately breaks when a future `0x03` `leadByte` is defined for the note plaintext.

Instead, we should store a boolean in `SaplingNote` indicating whether it is generated according to ZIP 212, and then only at serialization of a note plaintext do we interpret that to mean `leadByte = 0x02`. Then a future `0x03` that doesn't affect the note itself would only need to be handled at this interface, not every place that notes are generated.